### PR TITLE
[Merged by Bors] - Replace fixed timestep in `alien_cake_addict` example with timer

### PR DESCRIPTION
**Pull request by maniwani**
# Objective

Examples should use the correct tools for the job.

## Solution

A fixed timestep, by design, can step multiple times consecutively in a single update.

That property used to crash the `alien_cake_addict` example (#2525), which was "fixed" in #3411 (by just not panicking). The proper fix is to use a timer instead, since the system is supposed to spawn a cake every 5 seconds. 

---

A timer guarantees a minimum duration. A fixed timestep guarantees a fixed number of steps per second.
Each one works by essentially sacrificing the other's guarantee.

You can use them together, but no other systems are timestep-based in this example, so the timer is enough.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

File: examples/games/alien_cake_addict.rs [hunk 1]
```diff
@@ -9,9 +10,23 @@ enum GameState {
...
+struct SpawnTimer(Timer);
+
+fn timer_expired(time: Res<Time>, mut timer: ResMut<SpawnTimer>) -> ShouldRun {
+    timer.0.tick(time.delta());
+    if timer.0.finished() {
+        timer.0.reset();
+        ShouldRun::Yes
+    } else {
+        ShouldRun::No
+    }
```
Comment (devil-ira) [inline examples/games/alien_cake_addict.rs:23]: Instead of manually resetting this could be a repeating timer.
```suggestion
    timer.0.tick(time.delta()).just_finished().into()
```

File: examples/games/alien_cake_addict.rs [hunk 2]
```diff
@@ -9,9 +10,23 @@ enum GameState {
...
+}
+
 fn main() {
     App::new()
         .init_resource::<Game>()
+        .insert_resource(SpawnTimer(Timer::from_seconds(5.0, false)))
```
Comment (devil-ira) [inline examples/games/alien_cake_addict.rs:29]:
```suggestion
        .insert_resource(SpawnTimer(Timer::from_seconds(5.0, true)))
```

File: examples/games/alien_cake_addict.rs (hunk 1 above)
Comment (maniwani) [reply examples/games/alien_cake_addict.rs:23]: Do we actually want the wraparound behavior in this example?

If my game froze for 15 seconds and then resumed, would the cake not respawn three times in rapid succession?

Comment (Aceeri) [reply examples/games/alien_cake_addict.rs:23]: I think both of these changes will do the same thing just one is slightly more concise, since `just_finished` will be reset by the next `timer.0.tick(time.delta())`.

Comment (devil-ira) [reply examples/games/alien_cake_addict.rs:23]: In that case `just_finished` and `finished` would return true for that tick, and `times_finished_this_tick` would return `3`.

**Review by Weibye — CHANGES_REQUESTED** [review]
Changes look good but as @devil-ira points out, the timer should be repeating

**Review by JoJoJet — APPROVED** [review]

File: examples/games/alien_cake_addict.rs [hunk 3]
```diff
@@ -9,9 +10,23 @@ enum GameState {
     GameOver,
 }
 
+#[derive(Resource)]
+struct SpawnTimer(Timer);
+
```
Comment (cart) [inline examples/games/alien_cake_addict.rs:15]: This run criteria, in combination with the system set, feels like a lot of boilerplate for a behavior that is specific to the `spawn_bonus` behavior. In cases like this I think we should encourage the pattern of just checking the timer in the `spawn_bonus` system.

Comment (maniwani) [reply examples/games/alien_cake_addict.rs:15]: I think moving the check into the `spawn_bonus` system makes sense to simplify this example, but a condition/criteria is the best practice since we'll avoid spawning tasks when we don't need to.

The problem atm is we can't give `spawn_bonus` its own criteria and put it in the `Playing` state's `Update` system set, but we can clean that up once we've brought in all the stageless api changes.

File: examples/games/alien_cake_addict.rs (hunk 1 above)
Comment (maniwani) [reply examples/games/alien_cake_addict.rs:23]: I'm updating this to use the repeating timer.

Just wanted to leave a note that there are differences in behavior between a repeating timer and a timer that you manually reset to zero (no matter how much time elapsed).

The manually-reset timer "delays" future spawns while the repeating timer does not. But in the event of a super long duration between frames (i.e. massive hiccup), the repeating timer makes the "next-next" cake seem to spawn early.

**Comment by maniwani** [comment]
Updated to use repeating timer.

Just wanted to leave a note for future reference that there are differences in behavior between a repeating timer and a timer that you reset to zero whenever it's finished.

The manually-reset timer delays future spawns (if its elapsed time > the exact period). OTOH, in the event of a very long duration between frames (i.e. large hiccup), the repeating timer will make the "next-next" cake seem to spawn early.

File: examples/games/alien_cake_addict.rs [hunk 4]
```diff
@@ -291,13 +291,18 @@ fn focus_camera(
...
+    time: Res<Time>,
+    mut timer: ResMut<BonusSpawnTimer>,
     mut state: ResMut<State<GameState>>,
     mut commands: Commands,
     mut game: ResMut<Game>,
 ) {
-    if *state.current() != GameState::Playing {
+    // make sure we wait enough time before spawning the next cake
+    timer.0.tick(time.delta());
+    if !timer.0.finished() {
```
Comment (devil-ira) [inline examples/games/alien_cake_addict.rs:302]:
```suggestion
    if !timer.0.tick(time.delta()).finished() {
```
:)

**Comment by devil-ira** [comment]
> The manually-reset timer delays future spawns (if its elapsed time > the exact period). OTOH, in the event of a very long duration between frames (i.e. large hiccup), the repeating timer will make the "next-next" cake seem to spawn early.

Rather than the "next-next" cake spawning early, the problem would be the "next" cake spawning late because of the hiccup. A bit pedantic, I know, but that points to the real problem: the hiccup.
There is no way of making such a stutter pleasant. Even if we ignore the cakes spawning too close after one another, it will be incredibly jarring.
If he player wasn't limited to the grid in this game we would shoot probably off into the distance as well. The best thing is probably limiting the time delta globally.

A repeating timer not counting overflow has other consequences: it would drift.
Not too big a deal on its own, but different timers will then drift from eachother as well.

Let's say we added a time limit of 1 minute, so 60 seconds divided by 5 cakes per second is 12 cakes total.
If were ignoring overflow at 60 fps how much would it drift over those 11 finishes? probably not much, but what if a big hiccup happens? You could lose seconds which means you might only have 3-4 seconds to grab that last cake, instead of 5 :(

TL:DR
While ignoring overflow might make certain situations better, it's incorrect to do so and timers would start to drift.
Limiting the time delta globally might be a good idea, and we should look into supporting that, if not enabling it by default (probably after we get time-scaling in).

**Comment by maniwani** [comment]
> Rather than the "next-next" cake spawning early, the problem would be the "next" cake spawning late because of the hiccup. 
> You could lose seconds which means you might only have 3-4 seconds to grab that last cake, instead of 5.

Yeah, this is what I meant by "_seem_ to spawn early", was speaking from the perspective a player. You return from a hiccup, the next cake immediately spawns, but then the one after that will most likely spawn in before 5 seconds have passed.

> A repeating timer not counting overflow has other consequences: it would drift.

Yep, and this is why I agree a repeating timer as-is was the correct choice. I think big hiccups are mostly hypothetical and that I made a big deal out of them. It's good that we recognize that non-repeating (or non-wrapping ig) timers drift though.

> The best thing is probably limiting the time delta globally.

Clamping the `time.delta()` is definitely something we can do if time tracks both "app time" and "real-world time" (like in the global time scaling PR), but I dunno if we should yet. Need a few to think about how that would fit in with everything else.

**Review by alice-i-cecile — APPROVED** [review]

**Review by devil-ira — APPROVED** [review]
